examples/basic_auth.cpp

```diff
@@ -5,42 +5,67 @@
 #endif
 #include <stdlib.h>
 #include <signal.h>
-#include <mongoose/Server.h>
-#include <mongoose/WebController.h>
+#include <cassert>
+
+#include "Controller.h"
+#include "Request.h"
+#include "Response.h"
+#include "Server.h"
 
 using namespace std;
 using namespace Mongoose;
 
-class MyController : public WebController
+class MyController : public Controller
 {
-    public:
-        void hello(Request &request, StreamResponse &response)
-        {
-            response << "Hello " << htmlEntities(request.get("name", "... what's your name ?")) << endl;
-        }
-
-        void setup()
-        {
-            addRoute("GET", "/hello", MyController, hello);
-        }
+public:
+    bool hello(const std::shared_ptr<Request>& req, const std::shared_ptr<Response>& res)
```

Comment:
Is it possible that two handlers like this one gets the same response object? I mean is it possible that two threads are working on the same `Response *`? If so there's a problem because i saw response methods are not thread safe. If not why passing a shared pointer?

Reply:
Checked the rest of the code. Looks like shared pointers are everywhere. But I can't see what to share here. Looks like it should be `unique_ptr` or just rvalue. So we can transfer the ownership of request and response to other thread.

Reply:
@rferrazz nope. before "handling" any request, the Server.cpp creates a request/response pair.

@asergunov the shared pointer is mostly there to support "end users" of this library to do multithreaded things.

Like here: https://github.com/Proemion/quarks-recovery-server/blob/massive_cleanup/src/recoverycontroller.cpp#L135

Reply:
All the magic happens here:
https://github.com/Proemion/mongoose-cpp/blob/multithreaded-server/lib/Server.cpp#L43

That's where we create and destroy a request/response pair for each incoming connection. Then we simply use the controller to dispatch the request to the appropriate handler.

Reply:
I do have to test the multi threaded server a lot though. Thinking of writing unit tests for it later today.